sdk/core/core/src/az_http_request_builder.c

```diff
@@ -3,48 +3,201 @@
+        || new_url_size < appended_size || new_url_size < p_hrb->url.size
+        || new_url_size > p_hrb->max_url_size) {
+      return AZ_ERROR_BUFFER_OVERFLOW;
+    }
+
+    new_url_span.size = new_url_size;
+
+    // check whether name or value regions overlap with destination for some reason (unlikely)
+    if (az_const_span_is_overlap(az_span_to_const_span(new_url_span), name)
+        || az_const_span_is_overlap(az_span_to_const_span(new_url_span), value)) {
+      return AZ_ERROR_ARG;
+    }
+  }
+
+  // Find whether the URL contains '?'character already.
+  // So that we know if we should append "?text" or "&text".
+  // Scan from the end until the first occurrence of '?'.
+  bool first_parameter = true;
+  for (size_t i = p_hrb->url.size; i > 0;) {
+    --i;
+    if (p_hrb->url.begin[i] == '?') {
+      first_parameter = false;
+      break;
+    }
+  }
+
+  // Appent either '?' or '&'
+  p_hrb->url.begin[p_hrb->url.size] = first_parameter ? '?' : '&';
+  ++(p_hrb->url.size);
+
+  // Append parameter name
+  memcpy(p_hrb->url.begin + p_hrb->url.size, name.begin, name.size);
+  p_hrb->url.size += name.size;
+
+  // Append "=<parameter value>", if there's one.
+  if (has_value) {
+    p_hrb->url.begin[p_hrb->url.size] = '=';
+    ++(p_hrb->url.size);
+
+    memcpy(p_hrb->url.begin + p_hrb->url.size, value.begin, value.size);
+    p_hrb->url.size += value.size;
+  }
+
+  assert(p_hrb->url.size == new_url_size);
   return AZ_OK;
 }
 
 az_result az_http_request_builder_append_header(
     az_http_request_builder * const p_hrb,
     az_const_span const name,
     az_const_span const value) {
-  (void)p_hrb;
-  (void)name;
-  (void)value;
+  AZ_CONTRACT_ARG_NOT_NULL(p_hrb);
+
+  if (!az_const_span_is_valid(name) || !az_const_span_is_valid(value) || name.size == 0
+      || value.size == 0) {
+    return AZ_ERROR_ARG;
+  }
+
+  if (p_hrb->headers_end >= p_hrb->max_headers) {
+    return AZ_ERROR_BUFFER_OVERFLOW;
+  }
+
+  az_pair * const headers = get_headers_start(p_hrb->buffer, p_hrb->max_url_size);
+  headers[p_hrb->headers_end] = (az_pair){ .key = name, .value = value };
+  ++(p_hrb->headers_end);
+
   return AZ_OK;
 }
 
 az_result az_http_request_builder_mark_retry_headers_start(az_http_request_builder * const p_hrb) {
-  (void)p_hrb;
+  p_hrb->retry_headers_start = p_hrb->headers_end;
   return AZ_OK;
 }
 
 az_result az_http_request_builder_remove_retry_headers(az_http_request_builder * const p_hrb) {
-  (void)p_hrb;
+  if (p_hrb->headers_end > p_hrb->retry_headers_start) {
+    p_hrb->headers_end = p_hrb->retry_headers_start;
+  }
+
+  return AZ_OK;
+}
+
+az_result az_http_request_builder_get_header(
+    az_http_request_builder * const p_hrb,
+    uint16_t const index,
+    az_pair * const out_result) {
+  AZ_CONTRACT_ARG_NOT_NULL(p_hrb);
+  AZ_CONTRACT_ARG_NOT_NULL(out_result);
+
+  if (index >= p_hrb->headers_end) {
+    return AZ_ERROR_ARG;
+  }
+
+  az_pair * const headers = get_headers_start(p_hrb->buffer, p_hrb->max_url_size);
```

Comment:
It is a shame to recalc this when getting each and every header

Reply:
If you think it is serious, we can store az_pair* in the http_request_builder struct. I don't think it is perf critical. However, I like the idea of having az_pair* in the struct, so that users can iterate and read, the same way they can do with the url field. And I'd also ask for a separate buffer to store az_pair :) - we already have az_span_span type, see https://github.com/Azure/azure-sdk-for-c/blob/master/sdk/core/core/inc/az_span_seq.h

Reply:
I am keeping it as is, we can discuss more later. I am checking this in to master to unblock Victor.